### ISSUE TITLE
Unformatted blocks

### ISSUE THREAD

**Issue by andrewbaxter** [andrewbaxter/genemichaels#71]
There have been requests to only format certain things (code, not comments; code, not macros; just comments).  As a deterministic formatter genemichaels isn't built to consider original formatting at all, but it's something to consider if there's a way to do it that's sufficient and not a huge burden.

The main difficulties are:
* If the chunk shares lines with other code, it would be easy to go over the line limit since we wouldn't be to touch the chunk code
* For blacklisted chunks, indentation changes would somehow need to be applied, while keeping the same structure.  The subtree indentation could be estimated from the leading whitespace on the first line, unless the first line was shared with other code.  Similarly for whitelisted chunks, we'd need to somehow determine the indentation from surrounding text

This is a common pattern with partial formatting - rustfmt keeps growing indentation in split strings for example and you need to periodically go in and un-indent stuff manually.

**Comment by elfeiin** [comment]
So there needs to be a way to either peek what's nearby the input and do what genemichaels would do regardless of existing format, or initialize formatter state from the input based on heuristics. This is really a _selection_ formatting problem, which I don't think genemichaels supports. I don't know what metrics genemichaels tracks, but those could be read-in from the peeked code and then only mangling the procedural indentation to match adjacent lines depending on semantics.

To summarize, format everything at once, but don't clobber code outside the bounds of the selection. Keep source indentation of the adjacent lines if applicable, such as when formatting a selection or a comment within nested code.

**Comment by andrewbaxter** [comment]
To clarify, I think there are two cases here:
1. Formatting a file with genemichaels but skipping a chunk (generated code, or a very weird macro invocation, or something else that might need artisanal formatting)
1. Formatting a file with rustfmt but formatting modified lines with genemichaels

IIRC this issue and the use cases people mentioned that led to it were mostly about 1. I think 2 is unlikely to be widely useful (other developers undoing the formatting, project rustfmt linters rejecting it).

But as for heuristics (if you're curious), I think primarily the only information required based on the current formatting approach is the indentation of the nearest "wrapped" parent element, i.e. if you're formatting:
```
{
   {
      { a, b, some
.very
.very
.long
.expression() }
   }
}
```
to
```
{
   {
      { a, b, some
         .very
         .very
         .long
         .expression() }
   }
}
```
and not 
```
{
   {
      { a, b, some
            .very
            .very
            .long
            .expression() }
   }
}
```
That is, you'd want to know the indentation of the children of the first two `{` since they're wrapped, but not the 3rd `{` since it's not wrapped.  It might be reasonable to just assume every parent block is wrapped though.

If there's a good use case for 2 let's make it into a new issue.

**Comment by rootnas** [comment]
Can **genemichaels** format from stdin? It'd be nice to do `:pipe genemichaels`

**Comment by andrewbaxter** [comment]
Yeah! Right now it's `genemichaels --stdin` I think.

**Comment by andrewbaxter** [comment]
Just adding some more thoughts here:

I think the original use case for this is e.g. you're working on a change to someone else's repo and don't want to include extra lines in the diff (i.e. only format the lines you've modified).

I'm not sure this use case is valid:
- Genemichaels _should_ be deterministic, so if it was a genemichaels project, formatting everything shouldn't modify any lines you didn't change anyway
- If it's not a genemichaels project, I think maintainers probably wouldn't appreciate mixed styles and if they re-formatted it later using rustfmt it might introduce extra noise

I don't know if there are any other use cases.

Also, to expand on the difficulties of this:

The context for formatting isn't just the syntax subtree that it's formatting, but also the "indentation context" - where the indentation stops for various parent syntax elements are, which ones are active or not, etc.  This isn't something that can be obtained locally, or globally without knowing how the rest of the code was indented.  Getting it wrong could lead to stuff being unindented too far, indented too far, etc.

Ex:

```
    vec![vec![
        original element, original element, new element]
...
    ]
```

should this be wrapped as
```
    vec![vec![
        original element,
        original element,
        new element,
    ]
...
    ]
```

or 
```
    vec![vec![
        original element,
        original element,
        new element,
        ]
...
    ]
```

or 
```
    vec![vec![
            original element,
            original element,
            new element,
        ]
...
    ]
```

etc.  This is a small example but in more complex code the issue could be compounded and make something particularly unreadable I think.

**Comment by rootnas** [comment]
What if you format the whole file then only emit the selection?

**Comment by andrewbaxter** [comment]
Yeah, for internal formatting you could do that I think.  In fact, you could probably format just the subtree of the AST and adjust the starting indentation to match the surrounding or something? ...

**Comment by rootnas** [comment]
> Yeah, for internal formatting you could do that I think. In fact, you could probably format just the subtree of the AST and adjust the starting indentation to match the surrounding or something? ...

I wouldn't even bother... **genemichaels** is a deterministic formatter, so it should always produce the same output for a given input. You can source that determinism either by formatting the whole document and emitting only the selection, or by only formatting the selection. If a user is doing partial formatting, they are likely already prepared to indent anyway. But guess the indentation based on surrounding code breaks this determinism.
